text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+If an error occurs again, we can assume its proximate cause is in userland, not LWC.
+
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(MyApp, document.querySelector('#root')); 
+```
+
+### Constructable stylesheets
+
+The introduction of constructable stylesheets to LWC presents a small wrinkle. Since support for this browser API is not yet ubiquitous, we cannot know during SSR whether it will be supported on the client. For that reason, constructable stylesheets are disabled during SSR.
+
+From there, we have two options if support for constructable stylesheets is detected during rehydration:
+
+1. Swap in constructable stylesheets for vanilla stylesheets.
+2. Permanenly disable constructable stylesheets for the lifetime of the `document`.
+
+It is believed that swapping in constructable stylesheets could obviate the performance wins that constructable stylesheets are intended to bring. Furthermore, any theoretical benefit is outweighed by the implementation complexity of swapping in constructable stylesheets.
+
+For these reasons, it is proposed that constructable stylesheets be disabled for the lifetime of the `document` in a post-rehydration context.
+
+### Server-only components
+
+Addressing [Exception B to the happy path](#the-happy-path) will involve a solution roughly comparable to [React server components](https://reactjs.org/blog/2020/12/21/data-fetching-with-react-server-components.html). While the bulk of this can be implemented in user-land, at a level of abstraction above LWC, the `hydrateElement` patch implementation will need to support the use-case.
+
+However, we consider support for this case to be out-of-scope for the initial implementation. We will explore further as part of a separate RFC.
+
+### SSR placeholders
+
+Implementation of placeholder components, as described in [Exception A to the happy path](#the-happy-path) should require no extra effort, so long as unexpected subtree mismatches (Exception C) are dealt with properly.
+
+Implementation in user-land is possible with simple runtime environment detection:
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        return this.isNodeEnv ? tmplServerPlaceholder : tmplClient;

```

Comment:
I know this is pseudo-code, but I think it will be important to understand who is responsible for providing the host-specific environment setup. I do think we need to spec it out (plus its invariants - like should be immutable), and it should not be "user-land" defined.